File: 2025/02 February/24 Stock span problem/main.cpp
// Link: https://www.geeksforgeeks.org/problems/stock-span-problem-1587115621/1

#include <iostream>
#include <vector>
#include <stack>
using namespace std;

class Solution
{
public:
    vector<int> calculateSpan(vector<int> &arr)
    {
        int n = arr.size();

        vector<int> ans(n, 1);
        stack<int> s;

        for (int i = 0; i < n; i++)
        {
            while (!s.empty() && arr[s.top()] <= arr[i])
            {
                ans[i] += ans[s.top()];
                s.pop();
            }
            s.push(i);
        }
        return ans;
    }
};

int main()
{
    Solution object;
    vector<int> arr = {100, 80, 60, 70, 60, 75, 85};
    vector<int> ans = object.calculateSpan(arr);
    for (int i = 0; i < ans.size(); i++)
    {
        cout << ans[i] << " ";
    }
    cout << endl;

    arr = {10, 4, 5, 90, 120, 80};
    ans = object.calculateSpan(arr);
    for (int i = 0; i < ans.size(); i++)
    {
        cout << ans[i] << " ";
    }
    cout << endl;

    arr = {11, 4, 5, 90, 120, 80};
    ans = object.calculateSpan(arr);
    for (int i = 0; i < ans.size(); i++)
    {
        cout << ans[i] << " ";
    }
    cout << endl;

    return 0;
}

File: 2025/02 February/24 Stock span problem/main.py
# Link: https://www.geeksforgeeks.org/problems/stock-span-problem-1587115621/1


class Solution:
    def calculateSpan(self, arr):
        n = len(arr)
        ans = [1] * n
        s = []

        for i in range(n):
            while s and arr[s[-1]] <= arr[i]:
                ans[i] += ans[s.pop()]
            s.append(i)
        return ans


if __name__ == "__main__":
    object = Solution()
    arr = [100, 80, 60, 70, 60, 75, 85]
    ans = object.calculateSpan(arr)
    print(" ".join(map(str, ans)))

    arr = [10, 4, 5, 90, 120, 80]
    ans = object.calculateSpan(arr)
    print(" ".join(map(str, ans)))

    arr = [11, 4, 5, 90, 120, 80]
    ans = object.calculateSpan(arr)
    print(" ".join(map(str, ans)))
